### DISCUSSION TITLE
contactjs - v2.0

### DISCUSSION THREAD

**Discussion by biodiv** [biodiv/contactjs#16] [General]
Thanks to the great contrubutions of @SE2Dev the development of contactjs v2.0 is in progress.

### New layout of the source files
The TypeScript code is split into more specific .ts files. The architecture should become more readable and easier to understand.

### Introduction of PointerManager
The `Contact` class is replaced with the `PointerManager` class. `PointerManager` tracks added and removed `PointerInputs`. `PointerManager` also decides which type of Input currently is present: `SinglePointerInput`or `DualPointerInput`. Calcuations, previously partly done by `Contact` class are moved to a new dedicated `Geometry` class.

### Inputs are categorized more strictly.
The classes `SinglePointerInput` and `DualPointerInput` are introduced. Depending on the number of pointers active on the surface, only one type of PointerInput can be present. There cannot be two `SinglePointerInput`s at the same time. If two pointers are present, PointerManager interprets the situation always as `DualPointerInput`.

### Exact mapping of parameters
The collected, evaluated and required parameters, both `global` and `live` have the same naming. The concept of **intervals** and **boolean** parameters stays the same.

### No api changes
The api will stay the same.

**Comment by SE2Dev** [comment]
I want to point out that we should probably try to minimize the number of non-null assertions in the upgraded codebase. In the interest of keeping changes to the actual logic to a minimum, I had to add about 50 of them; I think that ultimately restructuring the code so that these aren't necessary would be a better approach.

There we also a few places where I was a bit iffy on the types, which could probably be cleaned up a bit. For example:

In `Gesture.emit()` there's a call to `this.getEventData()` which (according to the base `Gesture` class), returns a type that essentially looks like this:
```ts
interface GestureEventData {
  contact: Contact;
  recognizer: Gesture;
}
```

Shortly afterwards, however, there's an explicit lookup that assumes a `live` property to exist, even though some Gestures _may_ not necessarily have this property on their data types in the future.

```ts
// fire direction specific events
const currentDirection = eventData.live!.direction;
```

In practice, all exported `Gesture` types exported by the library _do_ add those properties, but we should probably come up with a more graceful solution as far as the actual types are concerned.

**Reply by biodiv** [reply]
I agree and tanks for sharing your insights. I already removed several possibilities for `null`. In some cases, like gesture parameters, the number value `0`should suffice for things like `speed` or `angle`. In theory of course, a _nonexistant angle_ represented by  _null_  is something different than an angle of 0 degrees, but I currently do not see a benefit of implementing the null concept here.

I did the same for `Vector`, which now if not present is a Vector of length 0.

I am not yet sure if `PointerListener.pointerManager` (formerly `PointerListener.contact`) should possibly be null. The former concept was, that a contact phenomenon starts an ends, and `PointerListener.contact` was set to null after the user removed the last pointer. This way it was ensured that data collected for pointers can be garbage collected if the contact instance reference was set to null.
Currently, `PointerListener.pointerManager` is an instance that lives as long as `PointerListener` lives. So in theory if a `Pointer` is not removed properly from `pointerManager`, it will not be garbage collected. There _should_ not be any dangling `Pointer` instances if the `pointercancel`and `pointerup` events are firing correctly.
